waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/server/FederatedHMSHandlerHive3.java

```diff
@@ -0,0 +1,674 @@
+import org.apache.hadoop.hive.metastore.api.WMCreateTriggerResponse;
+import org.apache.hadoop.hive.metastore.api.WMDropMappingRequest;
+import org.apache.hadoop.hive.metastore.api.WMDropMappingResponse;
+import org.apache.hadoop.hive.metastore.api.WMDropPoolRequest;
+import org.apache.hadoop.hive.metastore.api.WMDropPoolResponse;
+import org.apache.hadoop.hive.metastore.api.WMDropResourcePlanRequest;
+import org.apache.hadoop.hive.metastore.api.WMDropResourcePlanResponse;
+import org.apache.hadoop.hive.metastore.api.WMDropTriggerRequest;
+import org.apache.hadoop.hive.metastore.api.WMDropTriggerResponse;
+import org.apache.hadoop.hive.metastore.api.WMGetActiveResourcePlanRequest;
+import org.apache.hadoop.hive.metastore.api.WMGetActiveResourcePlanResponse;
+import org.apache.hadoop.hive.metastore.api.WMGetAllResourcePlanRequest;
+import org.apache.hadoop.hive.metastore.api.WMGetAllResourcePlanResponse;
+import org.apache.hadoop.hive.metastore.api.WMGetResourcePlanRequest;
+import org.apache.hadoop.hive.metastore.api.WMGetResourcePlanResponse;
+import org.apache.hadoop.hive.metastore.api.WMGetTriggersForResourePlanRequest;
+import org.apache.hadoop.hive.metastore.api.WMGetTriggersForResourePlanResponse;
+import org.apache.hadoop.hive.metastore.api.WMValidateResourcePlanRequest;
+import org.apache.hadoop.hive.metastore.api.WMValidateResourcePlanResponse;
+import org.apache.hadoop.hive.metastore.txn.TxnStore;
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.config.ConfigurableBeanFactory;
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Component;
+
+import com.jcabi.aspects.Loggable;
+
+import com.hotels.bdp.waggledance.mapping.service.MappingEventListener;
+import com.hotels.bdp.waggledance.mapping.service.impl.NotifyingFederationService;
+import com.hotels.bdp.waggledance.metrics.Monitored;
+
+@Monitored
+@Component
+@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+public class FederatedHMSHandlerHive3
+        extends FederatedHMSHandler
+{
+    private static final Logger LOG = LoggerFactory.getLogger(FederatedHMSHandlerHive3.class);
+
+    private AtomicInteger id = new AtomicInteger();
+
+    FederatedHMSHandlerHive3(MappingEventListener databaseMappingService, NotifyingFederationService notifyingFederationService)
+    {
+        super(databaseMappingService, notifyingFederationService);
+    }
+
+    @Override
+    @Loggable(value = Loggable.DEBUG, skipResult = true, name = INVOCATION_LOG_NAME)
+    public int getThreadId()
+    {
+        return id.incrementAndGet();
+    }
+
+    @Override
+    @Loggable(value = Loggable.DEBUG, skipResult = true, name = INVOCATION_LOG_NAME)
+    public RawStore getMS()
+            throws MetaException
+    {
+        throw new IllegalStateException("Called internally to metastore only so cannot be called by the proxy");
```

Comment:
Generally any API call that does not contain a database cannot be correctly proxied. We normally just  called the "primary" metastore for that. Reasoning is that most users use the primary for reading and writing and proxy to the federating ones just for reading which "generally" didn't involve these calls. 

That way at least users can do all the work on the primary via the proxy.

Any approach will have limitations but that has served us well in the past, so perhaps instead of throwing exception we delegate to the primary metastore.